text/2020-09-03-factored-prefix.md

```diff
@@ -0,0 +1,71 @@
+# Factored Key Prefix
+
+## Summary
+
+TiKV provides a facility where the client can set a key prefix that will be used for all keys in its requests.
+
+## Motivation
+
+TiKV users need to use key prefixes to avoid conflicts.
+In keeping with DRY we should not require a program to add the same prefix everywhere.
+This will also reduce the payload sent to TiKV.
+
+## Detailed design
+
+### Prefix Manipulation
+
+TiKV already adds a first byte prefix to all keys.
+TiKV adds a first byte prefix (currently 'z') to all data keys. Other prefixes are used for metadata.
+
+When the factored prefix is used, TiKV prepends a first byte prefix to the factored prefix sent by the client and prepends that to all keys that are sent.
+All keys are returned with the first byte prefix and factored prefix stripped.
+
+A factored prefix of 0x00 is not allowed. This helps avoid accidentally using null as a factored prefix.
+
+### API
+
+The factored prefix applies to all keys in a connection.
+So it is set as a GRPC header "TIKV-PREFIX" with the value set to the key prefix.
+
+### Backwards compatibility
```

Comment:
How about changes in PD?

Reply:
We don't require changes to PD. Over time we may want to further incorporate prefixes into our data model and have them be used by PD.